### Promise 의 상태
- 대기(pending) : 이행하거나 거부되지 않은 초기 상태
- 이행(fulfilled) : 연산이 성공적으로 완료됨.
- 거부(rejected) : 연산이 실패함

이행상태는 then으로 처리할 수 있다. resolve를 통해 전달한 값이 then에 인자로 전달된다.
### Promise 사용 방법!
```
const numbers = 10;
const promise = new Promise((resolve, reject) => {
//이 상태가 대기 fending 인 상태이다.
  if(numbers){ // 이행이 성공적으로 완료가 되면
    resolve('성공'); // resolve를 실행하게 된다.
  } else {		// 연산이 실패 하면
    reject('실패'); //reject를 실행한다.
  }
});
Promise.reject({errCode: 400})
  .catch(({errCode}) => console.log(errCode)) // 400
  거부상태는 catch으로 처리할 수 있다. reject를 통해 전달한 값이 catch에 인자로 전달된다.
```

### Promise.all() 사용  방법!
- Promise.all은 모두 이행상태일 때 then을 통해 결과를 받게 된다.
- 하나라도 거부상태가 되면 catch가 실행되게 된다.
```
Promise.all([
    getPromise(),
  ])
  .then(data => console.log(data))
  .catch(errCode => console.error(errCode))
```
### Await 정리!

- 프로미스 체이닝을 계속 하다보면 코드의 가독성이 떨어진다.
- async 와 await는 Promise를 간결/간편하고 동기적으로 실행되는것 처럼 보이게 만들어주는 API다.

### async&await 사용 방법!
```
const AsyncAwaitEx01 = async () => {
  
  // await는 async 함수 안에서만 동작합니다.
  let a = await getDataFromFilePromise(12345);
  let b = await getDataFromFilePromise(12345);
 //promise가 이행될 때까지 기다림
  
  return [JSON.parse(a) , JSON.parse(b)]; //완료!!
}
```

### fetch API를 사용하여 Promise를 사용했을 때 오류까지 throw 하는 방법
```
function logFetch(url) {
  return fetch(url)
    .then(response => response.text())
    .then(text => {
      console.log(text); // 성공!!
    }).catch(err => {
      console.error('fetch failed 404', err); //실패!
    });
}
```
### fetch API를 사용하여 async&await 사용했을 때 오류까지 예외처리(try catch) 하는 방법
```
async function logFetch(url) {
  try {
    const response = await fetch(url);
    console.log(await response.text());
  }
  catch (err) {
    console.log('fetch failed', err);
  }
}
```
## 프로그래머스 핸드폰 가리기

<div>문제 설명</div><br>

<div>프로그래머스 모바일은 개인정보 보호를 위해 고지서를 보낼 때 고객들의 전화번호의 일부를 가립니다.
전화번호가 문자열 phone_number로 주어졌을 때, 전화번호의 뒷 4자리를 제외한 나머지 숫자를 전부 *으로 가린 문자열을 리턴하는 함수, solution을 완성해주세요.</div>


|입출력|return|
|:---|---:
phone_number ||	return
"01033334444"||	"*******4444"
"027778888"	 || "*****8888"

<div>문제 풀이</div>

```
function solution(num) {
    const result = num.length - 4
    return "*".repeat(result) + num.substring(result)
}
```
<div>
1.정규표현식으로 하면 엄청 간단하게 풀 수 있는 문제지만 좀 더 생각을 해보면서 풀기로 결정을 했다.<br>
2.입력받은 값이 문자열이기 때문에 뒤에서 4번째까지 지정한값을 result에 할당한다.<br>
3.별찍기에서 사용했던 repeat메소드를 사용하여 *을 계속 찍기로 했고
substritng을 사용해 문자열을 result만큼 반환한다.<br>
</div>